modules/core/src/main/scala/doobie/util/read.scala

```diff
@@ -42,67 +44,120 @@ and similarly with Get:
 And find the missing instance and construct it as needed. Refer to Chapter 12
 of the book of doobie for more information.
 """)
-sealed abstract class Read[A](
-    val gets: List[(Get[?], NullabilityKnown)],
-    val unsafeGet: (ResultSet, Int) => A
-) {
-
-  final lazy val length: Int = gets.length
-
-  def map[B](f: A => B): Read[B] =
-    new Read(gets, (rs, n) => f(unsafeGet(rs, n))) {}
-
-  def ap[B](ff: Read[A => B]): Read[B] =
-    new Read(ff.gets ++ gets, (rs, n) => ff.unsafeGet(rs, n)(unsafeGet(rs, n + ff.length))) {}
+sealed trait Read[A] {
+  def unsafeGet(rs: ResultSet, startIdx: Int): A
+  def gets: List[(Get[?], NullabilityKnown)]
+  def toOpt: Read[Option[A]]
+  def length: Int
 
-  def get(n: Int): ResultSetIO[A] =
+  final def get(n: Int): ResultSetIO[A] =
     IFRS.raw(unsafeGet(_, n))
 
-}
+  final def map[B](f: A => B): Read[B] = Read.Composite[B](List(this), l => f(l.head.asInstanceOf[A]))
 
-object Read extends ReadPlatform {
+  final def ap[B](ff: Read[A => B]): Read[B] =
+    Read.Composite[B](List(ff, this), l => l(0).asInstanceOf[A => B](l(1).asInstanceOf[A]))
+}
 
-  def apply[A](
-      gets: List[(Get[?], NullabilityKnown)],
-      unsafeGet: (ResultSet, Int) => A
-  ): Read[A] = new Read(gets, unsafeGet) {}
+object Read extends LowerPriorityRead {
 
-  def apply[A](implicit ev: Read[A]): ev.type = ev
+  def apply[A](implicit ev: Read[A]): Read[A] = ev
 
-  def derived[A](implicit ev: MkRead[A]): Read[A] = ev
+  def derived[A](implicit
+      @implicitNotFound(
+        "Cannot derive Read instance. Please check that each field in the case class has a Read instance or can derive one")
+      ev: Derived[MkRead[A]]
+  ): Read[A] = ev.instance.underlying
 
-  trait Auto {
-    implicit def deriveRead[A](implicit ev: MkRead[A]): Read[A] = ev
-  }
+  trait Auto extends MkReadInstances
 
   implicit val ReadApply: Applicative[Read] =
     new Applicative[Read] {
       def ap[A, B](ff: Read[A => B])(fa: Read[A]): Read[B] = fa.ap(ff)
-      def pure[A](x: A): Read[A] = new Read(Nil, (_, _) => x) {}
+      def pure[A](x: A): Read[A] = Read.Composite(Nil, _ => x)
       override def map[A, B](fa: Read[A])(f: A => B): Read[B] = fa.map(f)
     }
 
-  implicit val unit: Read[Unit] =
-    Read(Nil, (_, _) => ())
+  implicit val unit: Read[Unit] = Read.Composite(Nil, _ => ())
+
+  implicit val optionUnit: Read[Option[Unit]] = Read.Composite(Nil, _ => Some(()))
 
-  implicit val optionUnit: Read[Option[Unit]] =
-    Read(Nil, (_, _) => Some(()))
+  implicit def fromReadOption[A](implicit read: Read[A]): Read[Option[A]] = read.toOpt
 
-  implicit def fromGet[A](implicit ev: Get[A]): Read[A] =
-    new Read(List((ev, NoNulls)), ev.unsafeGetNonNullable) {}
+  /** Simple instance wrapping a Put. i.e. single column non-null value */
+  final case class Single[A](get: Get[A]) extends Read[A] {
```

Comment:
Is there any need for this to be a `case class` instead of a `class`? It has no concept of equality and its not data.

Reply:
Thanks I've changed this for for both Read & Write